File: LeetCode/107_binary_tree_level_order_traversal_ii.py
""" november 2019
runtime: 32ms [74th percentile]
memory: 12.7Mb [?]
"""


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        if root is None:
            return
        
        
        nodeQueue = [root]
        res = []
        
        while len(nodeQueue) > 0:
            row = []
            
            for i in range(len(nodeQueue)):
                currNode = nodeQueue.pop()
                row.append(currNode.val)
                
                if currNode.left:
                    nodeQueue.insert(0, currNode.left)
                if currNode.right:
                    nodeQueue.insert(0, currNode.right)
                    
            res.append(row)
            
        
        return reversed(res)
            
            

File: LeetCode/107_binary_tree_level_order_traversal_ii.cpp
/* november 2019
runtime: 8ms [66th percentile]
memory: 13.7Mb 
*/


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#include <queue>

class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        if (!root)
            return {};
        vector<vector<int>> res;
        
        queue<TreeNode*> nodeQueue;
        nodeQueue.push(root);
        
        while (!nodeQueue.empty()){ // keep doing stuff if we have a lineup
            
            vector<int> currRow; // set up a vector to hold the current row
            int size= nodeQueue.size();
            
            for (int i=0; i<size; i++){ // iterate over the size of the queue
                TreeNode* currNode = nodeQueue.front();
                nodeQueue.pop();
                
                currRow.push_back(currNode->val);
                
                if (currNode->left)
                    nodeQueue.push(currNode->left);
                    
                if (currNode->right)
                    nodeQueue.push(currNode->right);
            }
            res.push_back(currRow);
        }
        
        reverse(res.begin(), res.end());
        return res;
    }
};